pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -200,9 +266,162 @@ func (meta *predicateMetadata) ShallowCopy() algorithm.PredicateMetadata {
+			return false
+		}
+	}
+	return true
+}
+
+// getPodsMatchingAffinity finds existing Pods that match affinity terms of the given "pod".
+// It ignores topology. It returns a set of Pods that are checked later by the affinity
+// predicate. With this set of pods available, the affinity predicate does not
+// need to check all the pods in the cluster.
+func getPodsMatchingAffinity(pod *v1.Pod, nodeInfoMap map[string]*schedulercache.NodeInfo) (affinityPods map[string][]*v1.Pod, antiAffinityPods map[string][]*v1.Pod, err error) {
+	allNodeNames := make([]string, 0, len(nodeInfoMap))
+
+	affinity := pod.Spec.Affinity
+	if affinity == nil || (affinity.PodAffinity == nil && affinity.PodAntiAffinity == nil) {
+		return nil, nil, nil
+	}
+
+	for name := range nodeInfoMap {
+		allNodeNames = append(allNodeNames, name)
+	}
+
+	var lock sync.Mutex
+	var firstError error
+	affinityPods = make(map[string][]*v1.Pod)
+	antiAffinityPods = make(map[string][]*v1.Pod)
+	appendResult := func(nodeName string, affPods, antiAffPods []*v1.Pod) {
+		lock.Lock()
+		defer lock.Unlock()
+		if len(affPods) > 0 {
+			affinityPods[nodeName] = affPods
+		}
+		if len(antiAffPods) > 0 {
+			antiAffinityPods[nodeName] = antiAffPods
+		}
+	}
+
+	catchError := func(err error) {
+		lock.Lock()
+		defer lock.Unlock()
+		if firstError == nil {
+			firstError = err
+		}
+	}
+
+	affinityProperties, err := getAffinityTermProperties(pod, GetPodAffinityTerms(affinity.PodAffinity))
+	if err != nil {
+		return nil, nil, err
+	}
+	antiAffinityProperties, err := getAffinityTermProperties(pod, GetPodAntiAffinityTerms(affinity.PodAntiAffinity))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	processNode := func(i int) {
+		nodeInfo := nodeInfoMap[allNodeNames[i]]
+		node := nodeInfo.Node()
+		if node == nil {
+			catchError(fmt.Errorf("nodeInfo.Node is nil"))
+			return
+		}
+		affPods := make([]*v1.Pod, 0, len(nodeInfo.Pods()))
+		antiAffPods := make([]*v1.Pod, 0, len(nodeInfo.Pods()))
+		for _, existingPod := range nodeInfo.Pods() {
+			// Check affinity properties.
+			if podMatchesAffinityTermProperties(existingPod, affinityProperties) {
+				affPods = append(affPods, existingPod)
+			}
+			// Check anti-affinity properties.
+			if podMatchesAffinityTermProperties(existingPod, antiAffinityProperties) {
+				antiAffPods = append(antiAffPods, existingPod)
+			}
+		}
+		if len(antiAffPods) > 0 || len(affPods) > 0 {
+			appendResult(node.Name, affPods, antiAffPods)
+		}
+	}
+	workqueue.Parallelize(16, len(allNodeNames), processNode)
+	return affinityPods, antiAffinityPods, firstError
+}
+
+// podMatchesAffinity returns true if "targetPod" matches any affinity rule of
+// "pod". Similar to getPodsMatchingAffinity, this function does not check topology.
+// So, whether the targetPod actually matches or not needs further checks for a specific
+// node.
+func targetPodMatchesAffinityOfPod(pod, targetPod *v1.Pod) bool {
```

Comment:
I think you could combine this function and `targetPodMatchesAntiaffinity` by having one more field for affinity or antiaffinity terms.

Reply:
We can, but the first half of the functions must be duplicated and more conditions checks must be added. I feel that would be less readable for not duplicating only several lines of code.

Reply:
SGTM